### PR TITLE
Adding asset containers & partials

### PR DESCRIPTION
**Pull request by robzienert**
Will allow for stuff like security group rules to be defined separately from the root security group.

### CODE REVIEW

File: keel-api/src/main/proto/Asset.proto [hunk 1]
```diff
@@ -14,6 +14,18 @@ message Asset {
   google.protobuf.Any spec = 4;
 }
 
+message PartialAsset {
+  AssetId id = 1;
```
Comment (robfletcher) [inline keel-api/src/main/proto/Asset.proto]: why does this have its own id?

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt [hunk 2]
```diff
@@ -5,11 +5,11 @@ import com.netflix.spinnaker.keel.api.plugin.CurrentResponse
 import io.grpc.stub.StreamObserver
 
 class TestAssetPlugin : AssetPluginGrpc.AssetPluginImplBase() {
-  override fun current(request: Asset, responseObserver: StreamObserver<CurrentResponse>) {
+  override fun current(request: AssetContainer, responseObserver: StreamObserver<CurrentResponse>) {
```
Comment (robfletcher) [inline keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt]: why does it need the partials in order to get the current state?

File: keel-api/src/main/proto/AssetPlugin.proto [hunk 3]
```diff
@@ -9,15 +9,15 @@ import "google/protobuf/empty.proto";
...
-  rpc plan (Asset) returns (PlanResponse);
+  rpc plan (AssetContainer) returns (PlanResponse);
 
-  rpc converge (Asset) returns (ConvergeResponse);
+  rpc converge (AssetContainer) returns (ConvergeResponse);
 }
 
 message CurrentResponse {
-  Asset asset = 1;
+  AssetContainer assetContainer = 1;
```
Comment (robfletcher) [inline keel-api/src/main/proto/AssetPlugin.proto]: surely the plugin is never going to respond with any partials

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt (hunk 2 above)
Comment (robfletcher) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt]: Feels like we should really just pass the `AssetId`.

File: keel-api/src/main/proto/Asset.proto (hunk 1 above)
Comment (robzienert) [reply keel-api/src/main/proto/Asset.proto]: It needs to be addressable. Partial assets will be saved / deleted independently from assets.

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt (hunk 2 above)
Comment (robzienert) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt]: Oh, here's why: Plugins won't know what to do with an `AssetId`. In the case of the AwsAssetPlugin, it's built entirely around the ability to inspect the request to decide what to do with it.

Comment (robfletcher) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt]: That's fair.

File: keel-api/src/main/proto/Asset.proto (hunk 1 above)
Comment (robfletcher) [reply keel-api/src/main/proto/Asset.proto]: Ok, I guess that makes sense.

File: keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt (hunk 2 above)
Comment (robfletcher) [reply keel-api/src/test/kotlin/com/netflix/spinnaker/keel/api/TestAssetPlugin.kt]: It doesn't need the partials, though. Right?

File: keel-rule-engine/src/main/kotlin/com/netflix/spinnaker/keel/persistence/AssetRepository.kt [hunk 4]
```diff
@@ -19,12 +22,23 @@ interface AssetRepository {
   fun get(id: AssetId): Asset?
 
   /**
-   * Persists an asset.
+   * Fetches the ids of any assets that depend (directly) on [id].
+   * Get a partial asset
    */
-  fun store(asset: Asset)
+  fun getPartial(id: AssetId): PartialAsset?
```
Comment (robfletcher) [inline keel-rule-engine/src/main/kotlin/com/netflix/spinnaker/keel/persistence/AssetRepository.kt]: Does it make sense to have a distinct id type?

Comment (robzienert) [reply keel-rule-engine/src/main/kotlin/com/netflix/spinnaker/keel/persistence/AssetRepository.kt]: You tell me. I don't think it matters, but I'm happy to add it in.

Comment (robfletcher) [reply keel-rule-engine/src/main/kotlin/com/netflix/spinnaker/keel/persistence/AssetRepository.kt]: I guess let's keep it the same for now & consider it separately from this PR.

**Comment by robzienert** [comment]
This is now perfect.

**Pull request merged**